File: kotlin/day07.kt
package day07
import kotlin.math.*
import common.readInputLines


fun median(nums: List<Double>): Double {
    var nc = nums.sorted()
    val mid = nc.size / 2
    if (nc.size % 2 == 1) {
        return nc[mid]
    } else {
        return (nc[mid - 1] + nc[mid]) / 2.0
    }
}

fun part1(nums: List<Double>): Int {
    val minx = round(median(nums)).toInt()
    return nums.map {abs(it - minx)}.sum().toInt()
}

fun minGradient(p0: Double, f: (Double) -> Double, num_it: Int): Double {
    val EPS = 0.001
    var p = p0
    for (i in 0 until num_it) {
        val pr = f(p + EPS)
        val pl = f(p - EPS)
        val dp = (pr - pl) / (2.0 * EPS)
        p = p - dp * EPS
    }
    return p
}

fun part2(nums: List<Double>): Int {
    val f = {x: Double ->
        var res = 0.0
        for (n in nums) {
            val d = abs(x - n)
            res += d * (d + 1.0) / 2.0
        }
        res
    };
    val minx = minGradient(median(nums), f, 1000)
    return f(round(minx)).toInt()
}

fun solution() {
    val line = readInputLines(7).first()
    val nums = line.split(",").map {it.toDouble()}
    println("Answer 1: ${part1(nums)}\nAnswer 2: ${part2(nums)}")
}

File: rust/src/day07.rs
use crate::common;

fn median(nums: &Vec<f64>) -> f64 {
    let mut nc = nums.clone();
    nc.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let mid = nc.len() / 2;
    if nc.len() % 2 == 1 {
        nc[mid]
    } else {
        (nc[mid - 1] + nc[mid]) / 2.0
    }
}

fn part1(nums: &Vec<f64>) -> i32 {
    let minx = median(nums).round();
    let sum: f64 = nums.iter().map(|n| (n - minx).abs()).sum();
    sum as i32
}

fn min_gradient(p0: f64, f: impl Fn(f64) -> f64, num_it: usize) -> f64 {
    let eps = 0.001;
    let mut p: f64 = p0;
    for _i in 0..num_it {
        let pr = f(p + eps);
        let pl = f(p - eps);
        let dp = (pr - pl) / (2.0 * eps);
        p = p - dp * eps;
    }
    p
}

fn part2(nums: &Vec<f64>) -> i32 {
    let f = |x: f64| {
        let mut res = 0.0;
        for n in nums {
            let d = (x - n).abs();
            res += d * (d + 1.0) / 2.0;
        }
        res
    };
    let minx = min_gradient(median(&nums), f, 1000).round();
    f(minx) as i32
}

pub(crate) fn solution() {
    let line: String = common::read_lines(&common::data_file(7)).unwrap().first().unwrap().clone();
    let nums: Vec<f64> = line.split(",").map(|x| x.parse::<f64>().unwrap()).collect();
    println!("Answer 1: {}\nAnswer 2: {}", part1(&nums), part2(&nums));
}
